[SVG] Add `async` attribute support in `SVGScriptElement`

Ahmad-S792 opened the pull request:
#### 5938e574e91a1610d37b0e038b2cb4141ed76f11
<pre>
[SVG] Add `async` attribute support in `SVGScriptElement`
<a href="https://bugs.webkit.org/show_bug.cgi?id=292835">https://bugs.webkit.org/show_bug.cgi?id=292835</a>
<a href="https://rdar.apple.com/problem/151561361">rdar://problem/151561361</a>

Reviewed by Anne van Kesteren.

Thia patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

This patch adds support for `async` attribute support in SVGScriptElement
similar to HTMLScriptElement as per web specification [1] &amp; [2]:

[1] <a href="https://www.w3.org/2012/01/05-svg-irc#T21-07-03">https://www.w3.org/2012/01/05-svg-irc#T21-07-03</a>
[2] <a href="https://www.w3.org/TR/SVG2/interact.html#ScriptElement">https://www.w3.org/TR/SVG2/interact.html#ScriptElement</a>

Following issue [3] is open to add it to SVG 2.0 IDL interfaces.

[3] <a href="https://github.com/w3c/svgwg/issues/918">https://github.com/w3c/svgwg/issues/918</a>

* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::attributeChanged):
(WebCore::SVGScriptElement::async const):
(WebCore::SVGScriptElement::hasAsyncAttribute const):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.idl:
* Source/WebCore/svg/svgattrs.in:
* LayoutTests/imported/w3c/web-platform-tests/svg/interact/scripted/async-01-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/svg/interact/scripted/async-04-expected.txt: Ditto
* LayoutTests/svg/dom/SVGScriptElement/script-async-attr-expected.txt: Removed.
* LayoutTests/svg/dom/SVGScriptElement/script-async-attr.svg: Removed.

Canonical link: <a href="https://commits.webkit.org/299735@main">https://commits.webkit.org/299735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1882c4e44ac87160c5ccde66d369c1fb450ade2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39682 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126319 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72062 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dd6da02a-df99-4180-a327-8b2f584c4a04) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121865 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91119 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60432 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/de286a46-a3a9-49d1-90b0-610ee1cd8af8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32252 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107593 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71674 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/31ce168f-4ab0-46d9-9471-3a4c4f518bb0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25698 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69956 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101727 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25886 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129239 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46909 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35571 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99737 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47275 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99582 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25288 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45038 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23057 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43530 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46771 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52477 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46237 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49586 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47923 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/svg/SVGScriptElement.cpp:
```diff
@@ -48,10 +49,29 @@ Ref<SVGScriptElement> SVGScriptElement::create(const QualifiedName& tagName, Doc
+void SVGScriptElement::setAsync(bool async)
+{
+    setBooleanAttribute(SVGNames::asyncAttr, async);
+    handleAsyncAttribute();
```
Doesn't this mean that `handleAsyncAttribute()` gets called twice?

Ahmad-S792 replied:
@annevk - it is to set `async` as false similar to HTMLScriptElement. From web specification - _When an [async](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-async) attribute is added to a [script](https://html.spec.whatwg.org/multipage/scripting.html#the-script-element) element el, the user agent must set el's [force async](https://html.spec.whatwg.org/multipage/scripting.html#script-force-async) to false._

annevk replied:
But that also happens as a result of the `async` attribute getting set, no? You have an attribute change handler above that also calls this.

Ahmad-S792 replied:
My interpretation of code is that:

`setBooleanAttribute` is setting value as `async` while next `handleAsyncAttribute` marks it as `m_forceAsync = false;` to ensure that the async is false.

Ahmad-S792 replied:
It is aligned with HTMLScriptElement here:

https://searchfox.org/wubkat/rev/d1661224f525bf15e34fde4eafe9de09b92c864b/Source/WebCore/html/HTMLScriptElement.cpp#186

```
void HTMLScriptElement::setAsync(bool async)
{
    setBooleanAttribute(asyncAttr, async);
    handleAsyncAttribute();
}
```

annevk replied:
I think that might also have the same effect of also calling `attributeChanged` and therefore `handleAsyncAttribute` getting called twice.

Ahmad-S792 replied:
If I remove the test still progresses but it is being done same way in HTMLScriptElement and in Blink as well (similar to our code). So I am hesitant to deviate away from what we are current doing. We probably could do better but rather do for both SVGScript and HTMLScript in one go as separate.

Ahmad-S792 replied:
@annevk - I just wanted to check on above to tackle both `HTMLScriptElement` and `SVGScriptElement` in follow-up to optimize in one go. Happy to file bug.

annevk replied:
That is in part what prevents us from using `[Reflect]` here and it also seems incorrect as `setBooleanAttribute` should call the attribute changed handler. It doesn't matter all that much I suppose given what `handleAsyncAttribute` ends up doing, but I thought you were interested in figuring out if `[Reflect]` could work? This would be part of that.

Ahmad-S792 replied:
I removed `[Reflect]` since `HTMLScriptElement` also didn't had it.

Ahmad-S792 replied:
@annevk - I think this is now ready. I rebased and used [ReflectSetter].

annevk commented on Source/WebCore/svg/SVGScriptElement.cpp:
```diff
@@ -48,10 +49,23 @@ Ref<SVGScriptElement> SVGScriptElement::create(const QualifiedName& tagName, Doc
 
+bool SVGScriptElement::async() const
+{
+    return hasAttributeWithoutSynchronization(SVGNames::asyncAttr) || forceAsync();
```
Maybe use `hasAsyncAttribute()` here?

Ahmad-S792 replied:
Done.

annevk reviewed (approved):
Would be nice to adopt ReflectSetter in HTMLScriptElement as well.

Ahmad-S792 committed:
[SVG] Add `async` attribute support in `SVGScriptElement`
https://bugs.webkit.org/show_bug.cgi?id=292835
rdar://problem/151561361

Reviewed by Anne van Kesteren.

Thia patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

This patch adds support for `async` attribute support in SVGScriptElement
similar to HTMLScriptElement as per web specification [1] & [2]:

[1] https://www.w3.org/2012/01/05-svg-irc#T21-07-03
[2] https://www.w3.org/TR/SVG2/interact.html#ScriptElement

Following issue [3] is open to add it to SVG 2.0 IDL interfaces.

[3] https://github.com/w3c/svgwg/issues/918

* Source/WebCore/svg/SVGScriptElement.cpp:
(WebCore::SVGScriptElement::attributeChanged):
(WebCore::SVGScriptElement::async const):
(WebCore::SVGScriptElement::hasAsyncAttribute const):
* Source/WebCore/svg/SVGScriptElement.h:
* Source/WebCore/svg/SVGScriptElement.idl:
* Source/WebCore/svg/svgattrs.in:
* LayoutTests/imported/w3c/web-platform-tests/svg/interact/scripted/async-01-expected.txt: Progression
* LayoutTests/imported/w3c/web-platform-tests/svg/interact/scripted/async-04-expected.txt: Ditto
* LayoutTests/svg/dom/SVGScriptElement/script-async-attr-expected.txt: Removed.
* LayoutTests/svg/dom/SVGScriptElement/script-async-attr.svg: Removed.

Canonical link: https://commits.webkit.org/299735@main
